kernel/posix/pthread.c

```diff
@@ -0,0 +1,550 @@
+ *
+ * See IEEE 1003.1
+ */
+int pthread_attr_setschedparam(pthread_attr_t *attr,
+			       const struct sched_param *schedparam)
+{
+	int priority = schedparam->priority;
+
+	if (!attr || !attr->initialized) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (is_posix_prio_valid(priority, attr->schedpolicy) == false) {
+		errno = ENOTSUP;
+		return -1;
+	}
+
+	attr->priority = priority;
+	return 0;
+}
+
+/**
+ * @brief Set stack attributes in thread attributes object.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_attr_setstack(pthread_attr_t *attr, void *stackaddr,
+			  size_t stacksize)
+{
+	if (stackaddr == NULL) {
+		errno = EACCES;
+		return -1;
+	}
+
+	attr->stack = stackaddr;
+	attr->stacksize = stacksize;
+	return 0;
+}
+
+static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+{
+	void * (*fun_ptr)(void *);
+
+	fun_ptr = arg3;
+	(*fun_ptr)((void *)arg1);
+}
+
+/**
+ * @brief Create a new thread.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
+		   void *(*threadroutine)(void *), void *arg)
+{
+	static atomic_val_t thread_count;
+	atomic_val_t thread_index = 0;
+	s32_t prio;
+	pthread_mutexattr_t att;
+	pthread_condattr_t cond_attr;
+	struct posix_thread *thread;
+
+	if (!attr || !attr->initialized || !attr->stack || !attr->stacksize) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	prio = posix_to_zephyr_priority(attr->priority, attr->schedpolicy);
+	thread_index = atomic_inc(&thread_count);
+
+	/* FIXME: Thread space allocation need to use mempool for better
+	 * space utilization under limited number of threads.
+	 */
+	if (thread_index < CONFIG_MAX_PTHREAD_COUNT) {
+		thread = &pthread[thread_index];
+		thread->cancel_state = (1 << PTHREAD_CANCEL_POS) & attr->flags;
+		thread->state = attr->detachstate;
+		thread->cancel_pending = 0;
+		thread->state_lock.sem = &thread->state_lock_sem;
+		pthread_mutex_init(&thread->state_lock, &att);
+		thread->cancel_lock.sem = &thread->cancel_lock_sem;
+		pthread_mutex_init(&thread->cancel_lock, &att);
+		pthread_cond_init(&thread->state_cond, &cond_attr);
+
+		*newthread = (pthread_t)
+				k_thread_create(&thread->thread,
+						attr->stack, attr->stacksize,
+						(k_thread_entry_t)
+						zephyr_thread_wrapper,
+						(void *)arg, NULL,
+						threadroutine, prio,
+						attr->flags,
+						attr->delayedstart);
+
+	} else {
+		/* Insuffecient resource, maximum thread count reached */
+		errno = EAGAIN;
+		return -1;
+	}
+
+	return 0;
+}
+
+
+/**
+ * @brief Set cancelability State.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_setcancelstate(int state, int *oldstate)
+{
+	struct posix_thread *pthread = (struct posix_thread *) pthread_self();
+
+	if (state != PTHREAD_CANCEL_ENABLE &&
+	    state != PTHREAD_CANCEL_DISABLE) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	*oldstate = pthread->cancel_state;
+	pthread_mutex_lock(&pthread->cancel_lock);
+	pthread->cancel_state = state;
+
+	if (state == PTHREAD_CANCEL_DISABLE) {
+		_thread_essential_set();
```

Comment:
While using the cancel_lock mutex is fine, saying that a thread is essential is only possible in supervisor mode. While this is fine, it's one of the things that must be documented.

Reply:
So what is your suggestion, Shall we stop propagating the THREAD_CANCEL_ENABLE/DISABLE calls to zephyr and handle it in POSIX layer only ? 
Also when you mentioned documentation - Shall we document inline to code only or you suggest to create documentation separately and put it under "doc" directory in Zephyr.

Reply:
It's fine to keep it this way. Only document that this function won't work in user mode. We can revisit this at a later date.